Game name autocompletion added to search bar

ptlin25 opened the pull request:
The search bar now autocompletes game names when the Game option is selected in the dropdown menu. The search bar finds the first 10 autocompletion results and displays the tags underneath the search bar. Clicking on the tag will redirect the user to the appropriate page. If a different search option is selected, the autocompletion is disabled for now. This can be changed in the future to autocomplete the relevant tags.

Commit:
Initial attempt at autocomplete

Commit:
Merge branch 'dev' into game-management/autocomplete

Commit:
Fixed format issue caused by merge

Commit:
Format has errors

Commit:
Merge branch 'dev' into game-management/autocomplete

Commit:
Merge branch 'dev' into game-management/autocomplete

Commit:
Added djlint tags

Commit:
Autocomplete changes when query type changes

Commit:
Merge branch 'dev' into game-management/autocomplete

Commit:
Only show top 10 autocomplete results

Commit:
Merge branch 'dev' into game-management/autocomplete

rakrusch commented on src/templates/base.html:
```diff
@@ -136,23 +142,91 @@
+                  return matcher.test(tag.label)
+                })
+                // only display the first 10 results
+                response(results.slice(0, 10));
```
Minor unnecessary stylistic idea, but we could consider creating a constant that we set equal to 10 and have that constant here to improve readability and make it easier if we ever want to change the number of results we are displaying.

majorsylvie replied:
yes I agree 100%, avoiding magic numbers and instead having this be stored as some `AUTOCOMPLETION_RESULTS_LIMIT` constant.

rakrusch commented on src/templates/base.html:
```diff
@@ -136,23 +142,91 @@
 
-          query_type.addEventListener("change", function() {
+          // by default, query autocompletes game names
+          {% get_games as games %}
```
I really like your use of comments to explain the purpose of code! If you want to more detailed with comments, you could related to important logical steps. For example, here you could add a comment next to 
{% get_games as games %}
 stating that we are:
// Fetching game names for autocomplete

Though, this is not strictly necessary and your use of comments already is really good.

majorsylvie replied:
I agree, adding an explicit comment saying line 182 is a query would be quite nice.

rakrusch commented on src/templates/base.html:
```diff
@@ -136,23 +142,91 @@
+                      return matcher.test(tag.label)
+                    })
+                    // only display the first 10 results
+                    response(results.slice(0, 10));
```
Same idea here with using a constant

rakrusch reviewed (approved):
This looks really good Patrick! I really like your use of inline comments to make code much more readable. I made some small comments but will approve this PR anyway, as the changes I suggest are minor and not strictly necessary.

ptlin25 committed:
Merge branch 'dev' into game-management/autocomplete

majorsylvie committed:
Merge branch 'dev' into game-management/autocomplete

majorsylvie commented on src/chigame/games/templatetags/tags.py:
```diff
@@ -0,0 +1,11 @@
+
+from chigame.games.models import Game
+
+register = template.Library()
```
Please add comments explaining what is happening in this code here. 
As template tags are something not covered in the course IIRC, thus explaining more of this code would be good for the rest of the chigame devs.

majorsylvie replied:
great comment added!

majorsylvie commented on src/chigame/games/templatetags/tags.py:
```diff
@@ -0,0 +1,11 @@
+
+
+@register.simple_tag
+def get_games():
```
Similar to above, add a comment explaining why this code is here and what purpose it serves.

majorsylvie commented on src/templates/base.html:
```diff
@@ -171,17 +178,83 @@
+                /*
+                new regular expression object (RegExp) to search for pattern
+                request.term is the string entered into the search bar
+                "i" means that the matching is case insensitive
```
great inclusion with this `'i'` comment

majorsylvie commented on src/templates/base.html:
```diff
@@ -171,17 +178,83 @@
-    {% endblock inline_javascript %}
-  </body>
+</script><!-- djlint:on -->
+{# turn djlint back on#}
```
I would add a comment about why this is necessary!

majorsylvie reviewed (changes requested):
The functionality of no-autocomplete while selecting the `Users` from dropdown does not appear to work:

https://github.com/uchicago-cs/chigame/assets/86740254/16a285c0-3337-4d6c-8f69-39ec333c360d


Besides that, the actual autocompletion is wonderful and look genuinely good. I appreciate that you are directly linked to the page and the choice to limit to the top 10 results. 

Please add the code comments and constant, then actually make sure that you're dropdown-switching works

Commit:
Added comments and fixed autocomplete when users is selected

Commit:
Further comments and documentation

ptlin25 commented:
Thanks for catching the issue with the autocompletion of game names even when the user option is selected. I have fixed that issue and also added more comments to explain the code.

ptlin25 committed:
Merge branch 'dev' into game-management/autocomplete

ptlin25 committed:
Merge branch 'dev' into game-management/autocomplete

majorsylvie committed:
Merge branch 'dev' into game-management/autocomplete

ptlin25 committed:
Merge branch 'dev' into game-management/autocomplete

Commit:
Fixing linter issues

Commit:
query changed to query-input

majorsylvie reviewed (approved):
Now with the forum dropdown merged in #309, this code still works and only autocompletes for the games search!

Great work!

One thing to note, you may want to consider some kind of search ranking that prioritizes the beginnings of words / titles.
Since here searching with `"chi"` returns `Luftschiff` before `Chicago`, which seems odd.

Doesn't have to be a change made in this PR, **please add an issue to the backlog to do this if you feel it's possible!**
<img width="485" alt="image" src="https://github.com/uchicago-cs/chigame/assets/86740254/d9dd79b8-adea-4e51-a91e-31978e115563">

ptlin25 commented:
When the search bar in the navigation bar has the "Games" option in the dropdown menu selected, the search bar will autocomplete game names in the database. This is done through the jQuery UI Autocomplete widget. A custom template tag was added to get a list of all of the games in the chigame database. This list is used to create the autocomplete tags. After a user enters a phrase into the search bar, the first 10 autocomplete results are shown. If the user clicks on the autocomplete options, they will be redirected to the corresponding game's detail page. Currently, there is no autocomplete for other search bar modes. More info can be found on the [wiki page](https://github.com/uchicago-cs/chigame/wiki/Games-~-Game-Name-Autocompletion).

majorsylvie commented:
Issue Score: Excellent

Comments:
Phenomenal work! This is an awesome PR that was very thoroughly done. 
I appreciate the great closing comments, good reviews (thanks @rakrusch)

Be very proud of this great work! We're almost done with the quarter!

<!-- {"score": "excellent", "minor": false} -->
